sympy/matrices/matrices.py

```diff
@@ -3912,35 +3975,121 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
         LUdecompositionFF
         LUsolve
         """
-        if not self.is_square:
-            raise NonSquareMatrixError(
-                "A Matrix must be square to apply LUdecomposition_Simple().")
-        n = self.rows
-        A = self.as_mutable()
-        p = []
-        # factorization
-        for j in range(n):
-            for i in range(j):
-                for k in range(i):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-            pivot = -1
-            for i in range(j, n):
-                for k in range(j):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-                # find the first non-zero pivot, includes any expression
-                if pivot == -1 and not iszerofunc(A[i, j]):
-                    pivot = i
-            if pivot < 0:
-                # this result is based on iszerofunc's analysis of the possible pivots, so even though
-                # the element may not be strictly zero, the supplied iszerofunc's evaluation gave True
-                raise ValueError("No nonzero pivot found; inversion failed.")
-            if pivot != j:  # row must be swapped
-                A.row_swap(pivot, j)
-                p.append([pivot, j])
-            scale = 1 / A[j, j]
-            for i in range(j + 1, n):
-                A[i, j] = A[i, j] * scale
-        return A, p
+
+        if rankcheck:
+            SymPyDeprecationWarning(
+                feature="Keyword argument rankcheck is deprecated. "
+                        "Future LUdecomposition_Simple() will complete "
+                        "regardless rank of input matrix.",
+                issue=9796,
+                deprecated_since_version="1.1").warn()
+
+        if self.rows == 0 or self.cols == 0:
+            # Define LU decomposition of a matrix with no entries as a matrix
+            # of the same dimensions with all zero entries.
+            return self.zeros(self.rows, self.cols), []
+
+        lu = self.as_mutable()
+        row_swaps = []
+
+        pivot_col = 0
+        for pivot_row in range(0, lu.rows-1):
+            # Search for pivot. Prefer entry that iszeropivot determines
+            # is nonzero, over entry that iszeropivot cannot guarantee
+            # is  zero.
+            candidate_pivot_row = None
+            while pivot_col != self.cols:
+                for r in range(pivot_row, self.rows):
+                    iszeropivot = iszerofunc(lu[r, pivot_col])
```

Comment:
Looks like I was doing it wrong:
```
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_on')
>>> p.sort_stats('cumulative').print_stats(10)
Thu Mar 30 18:43:46 2017    /tmp/find_reasonable_pivot_on

         384704 function calls (377602 primitive calls) in 2.781 seconds

   Ordered by: cumulative time
   List reduced from 2116 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    2.796    2.796 find_reasonable_pivot_test.py:1(<module>)
        1    0.005    0.005    1.983    1.983 sympy/__init__.py:12(<module>)
        1    0.026    0.026    0.885    0.885 sympy/polys/__init__.py:1(<module>)
       10    0.001    0.000    0.812    0.081 find_reasonable_pivot_test.py:3(main)
        1    0.005    0.005    0.669    0.669 sympy/core/__init__.py:2(<module>)
       10    0.000    0.000    0.573    0.057 sympy/printing/pretty/pretty.py:2162(pretty_print)
       10    0.058    0.006    0.538    0.054 sympy/printing/pretty/pretty.py:2144(pretty)
       10    0.000    0.000    0.480    0.048 sympy/printing/pretty/pretty.py:61(doprint)
   449/10    0.003    0.000    0.476    0.048 sympy/printing/printer.py:235(_print)
       10    0.000    0.000    0.476    0.048 sympy/printing/pretty/pretty.py:664(_print_MatrixBase)


<pstats.Stats instance at 0x10c447c68>
```

```
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_off')
>>> p.sort_stats('cumulative').print_stats(10)
Thu Mar 30 18:44:09 2017    /tmp/find_reasonable_pivot_off

         373177 function calls (366716 primitive calls) in 2.467 seconds

   Ordered by: cumulative time
   List reduced from 2108 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    2.522    2.522 find_reasonable_pivot_test.py:1(<module>)
        1    0.027    0.027    1.900    1.900 sympy/__init__.py:12(<module>)
        1    0.003    0.003    0.670    0.670 sympy/core/__init__.py:2(<module>)
        1    0.014    0.014    0.628    0.628 sympy/polys/__init__.py:1(<module>)
       10    0.001    0.000    0.621    0.062 find_reasonable_pivot_test.py:3(main)
       10    0.000    0.000    0.426    0.043 sympy/printing/pretty/pretty.py:2162(pretty_print)
       10    0.000    0.000    0.395    0.039 sympy/printing/pretty/pretty.py:2144(pretty)
       10    0.000    0.000    0.394    0.039 sympy/printing/pretty/pretty.py:61(doprint)
   449/10    0.003    0.000    0.391    0.039 sympy/printing/printer.py:235(_print)
       10    0.000    0.000    0.391    0.039 sympy/printing/pretty/pretty.py:664(_print_MatrixBase)


<pstats.Stats instance at 0x10c299d40>
```
I'll base the pivot search method on `_find_reasonable_pivot()`. As long as the caller of `LUdecomposition_simple()` can turn off unwanted simplifications when it "knows" it doesn't need them, then I have no objection to using ` _find_reasonable_pivot()`.